sklearn/model_selection/_subselect.py

```diff
@@ -0,0 +1,573 @@
+    >>> from sklearn.decomposition import PCA
+    >>> from sklearn.svm import SVC
+    >>> from sklearn.pipeline import Pipeline
+    >>> from sklearn.model_selection import by_standard_error
+    >>> X, y = load_digits(return_X_y=True)
+    >>> pipe = Pipeline([
+            ("reduce_dim", PCA(random_state=42)),
+            ("classify", SVC(random_state=42, C=0.01))
+        ])
+    >>> param_grid = {"reduce_dim__n_components": [6, 8, 10, 12, 14]}
+    >>> search = GridSearchCV(pipe, param_grid=param_grid, scoring="accuracy")
+    >>> search.fit(X, y)
+    >>> ss = Refitter(search.cv_results_)
+    >>> ss.fit(by_standard_error(sigma=1))
+    >>> refitted_index = ss.transform("reduce_dim__n_components")
+    >>> refitted_index
+
+    """
+
+    def __init__(self, cv_results_: Dict, scoring: str = "score"):
+        self.cv_results_ = cv_results_
+        self.scoring = scoring
+
+    def _get_splits(self):
+        # Extract subgrid corresponding to the scoring metric of interest
+        fitted_key_strings = "\t".join(list(self.cv_results_.keys()))
+        if not all(s in fitted_key_strings for s in ["split", "params", "mean_test"]):
+            raise TypeError(
+                "cv_results_ must be a dict of fitted GridSearchCV or RandomSearchCV"
+                " objects."
+            )
+
+        _splits = [
+            i
+            for i in list(self.cv_results_.keys())
+            if f"test_{self.scoring}" in i and i.startswith("split")
+        ]
+        if len(_splits) == 0:
+            raise KeyError(f"Scoring metric {self.scoring} not found in cv grid.")
+        else:
+            return _splits
+
+    @property
+    def _n_folds(self):
+        # Extract number of folds from cv_results_. Note that we cannot get this from
+        # the `n_splits_` attribute of the `cv` object because it is not exposed to the
+        # refit callable.
+        return len(
+            list(
+                set(
+                    [
+                        i.split("_")[0]
+                        for i in list(self.cv_results_.keys())
+                        if i.startswith("split")
+                    ]
+                )
+            )
+        )
+
+    @property
+    def _score_grid(self):
+        # Extract subgrid corresponding to the scoring metric of interest
+        return np.vstack([self.cv_results_[cv] for cv in self._get_splits()]).T
+
+    @property
+    def _cv_means(self):
+        # Calculate means of subgrid corresponding to the scoring metric of interest
+        return np.array(np.nanmean(self._score_grid, axis=1))
+
+    @property
+    def _lowest_score_idx(self):
+        # Return index of the lowest performing model
+        return np.nanargmin(self._cv_means)
+
+    @property
+    def _best_score_idx(self):
+        # Return index of the highest performing model
+        return np.nanargmax(self._cv_means)
+
+    def _apply_thresh(self, param: str, l_cutoff: float, h_cutoff: float) -> int:
```

Comment:
I don't see how this method, or any other method in `Refitter` uses the values of `param`. Why do we need this to be a functon of `param` when it simply selects the lowest scoring model among those within the range.

Reply:
Does the range here always include the highest score? If so, is there any need for `h_cutoff`? Are we not just needing a `min_score` value?

Regardless, when it comes to thresholding, I prefer parameters named `min_*` or `max_*` since these make it clear what the correct behaviour should be in case of equality.

Reply:
* I've modified things across the board to now make `param` an optional argument, which I think turned out to nicely generalize Refitter's functionality further.  

* Thresholding variables have now been renamed to `min_cut` and `max_cut`.

Reply:
Because you're not limiting the user in any case: it's not so hard to define your own function, especially for a predetermined range. But it does add to the maintenance burden, in terms of time taken to understand the code, to ensure it is correct and to modify it.

Reply:
I've not looked at the most recent changes, but I'm not sure why even an optimal param is useful. And hence why do we need `constrain` as separate from the individual refit callable factories?

Reply:
If I'm understanding correctly, it sounds like there are two separate issues being raised here:
1) the `constrain` function adds to the maintenance burden because it appears to have some redundancy with `Refitter`. 
_Response:_
There is indeed some overlap, but they actually address distinct use-cases. In its current implementation, `Refitter` alone can only be fit to `cv_results_` post-hoc (i.e. after a `SearchCV` has been fit) and _not_ as an argument to `refit=`. `constrain` (formerly the `simplify` @staticmethod of `Razors`) is no longer part of a unified design with the `Refitter` class; it now lives as its own independent API. It does what `Refitter` alone cannot-- instantiates a function-partial callable that wraps `Refitter` under the hood, but which can be directly supplied to the `refit=` argument by `SearchCV` _prior_ to fit.  I've added examples in the docstring for both `Refitter` and `constrain` that hopefully help to highlight this distinction for the user and demonstrate their different use-cases in practice. If you feel that there is still ambiguity on this or more documentation is needed, however, please don't hesitate to let me know and I'll add it. Please also let me know you think it'd be better for us to re-attach `constrain` as a @staticmethod to `Refitter`, because we can certainly still do that too if the unified design is indeed preferable for this aspect specifically?

2) the `param` argument is unnecessary, because it simply selects the lowest scoring model among those within the range determined by a given selector. 
_Response:_
The original motivation for including a `param` argument in `Refitter` would not be to select the lowest scoring model, but to select the simplest one that is not significant different from the highest-scoring one. With that said, I do agree that having additional logic around `param` (even as an optional argument) may risk creating confusion at first, and add more cognitive complexity to this PR than is needed, so I've gone ahead and removed `param` altogether for now i.e. "the simplest pull request is the best one"  (pun).

Let me know if I missed anything around these issues, or if there are other changes still needed.

Hope you had a nice holiday, and happy 2023!